src/java.base/share/classes/com/sun/crypto/provider/PBMAC1Parameters.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * algid-hmacWithSHA1 AlgorithmIdentifier {{PBKDF2-PRFs}} ::=
+ *     {algorithm id-hmacWithSHA1, parameters NULL : NULL}
+ *
+ * id-hmacWithSHA1 OBJECT IDENTIFIER ::= {digestAlgorithm 7}
+ *
+ * </pre>
+ *
+ * @since 26
+ */
+abstract class PBMAC1Parameters extends AlgorithmParametersSpi {
+
+    private static final ObjectIdentifier pkcs5PBKDF2_OID =
+            ObjectIdentifier.of(KnownOIDs.PBKDF2WithHmacSHA1);
+
+    // the PBMAC1 algorithm name
+    private String pbmac1AlgorithmName = null;
+
+    // the salt
+    private byte[] salt = null;
+
+    // the iteration count
+    private int iCount = 0;
+
+    // the Hmac function used by the kdf
+    private String kdfHmac = null;
+
+    // the Hmac function
+    private String Hmac = null;
+
+    // the key derivation function (default is HmacSHA1)
+    private final ObjectIdentifier kdfAlgo_OID =
+            ObjectIdentifier.of(KnownOIDs.HmacSHA1);
+
+    // length of key generated by the kdf
+    private int keyLength = -1;
+
+    protected void engineInit(AlgorithmParameterSpec paramSpec)
+        throws InvalidParameterSpecException
+    {
+       if (!(paramSpec instanceof PBMAC1ParameterSpec)) {
+           throw new InvalidParameterSpecException
+               ("Inappropriate parameter specification");
+       }
+       salt = ((PBMAC1ParameterSpec)paramSpec).getSalt().clone();
+       iCount = ((PBMAC1ParameterSpec)paramSpec).getIterationCount();
+    }
+
+    protected void engineInit(byte[] encoded)
+        throws IOException
+    {
+        DerValue pBMAC1_params = new DerValue(encoded);
+        if (pBMAC1_params.tag != DerValue.tag_Sequence) {
+            throw new IOException("PMAC1 parameter parsing error: "
+                + "not an ASN.1 SEQUENCE tag");
+        }
+        DerValue[] Info = (new DerInputStream(pBMAC1_params.toByteArray()))
+                .getSequence(2);
+        if (Info.length != 2) {
+            throw new IOException("PMAC1 parameter parsing error: "
+                + "expected length not 2");
+        }
+        ObjectIdentifier OID = Info[1].data.getOID();
+            KnownOIDs o = KnownOIDs.findMatch(OID.toString());
+            if (o == null || (!o.stdName().equals("HmacSHA1") &&
+                    !o.stdName().equals("HmacSHA224") &&
+                    !o.stdName().equals("HmacSHA256") &&
+                    !o.stdName().equals("HmacSHA384") &&
+                    !o.stdName().equals("HmacSHA512") &&
+                    !o.stdName().equals("HmacSHA512/224") &&
+                    !o.stdName().equals("HmacSHA512/256"))) {
+                throw new IOException("PBMAC1 parameter parsing error: "
+                        + "expecting the object identifier for a HmacSHA key "
+                        + "derivation function");
+            }
+            Hmac = o.stdName();
+
+        DerValue kdf = pBMAC1_params.data.getDerValue();
+        var kdfParams = new PBKDF2Parameters();
+        String kdfAlgo = kdfParams.parseKDF(kdf);
```

Comment:
Maybe just use a `PBKDF2Parameters(DerValue) `constructor? Then retrieve the algorithm via a separate getter method.

Reply:
That constructor doesn't exist. I don't know how I would do what you suggest.